Undefined symbols ___cpuid_count ___get_cpuid_max building on Mac OS X 10.7

htslib 1.20 fails to build on Mac OS X 10.7.5 with the version of Apple's fork of the clang compiler (`Apple LLVM version 4.2 (clang-425.0.28) (based on LLVM 3.2svn)`) that comes with Xcode 4.6.3 (the last version compatible with Mac OS X 10.7) due to:

```
Undefined symbols for architecture x86_64:
  "___cpuid_count", referenced from:
      _htscodecs_tls_cpu_init in rANS_static4x16pr.o
  "___get_cpuid_max", referenced from:
      _htscodecs_tls_cpu_init in rANS_static4x16pr.o
ld: symbol(s) not found for architecture x86_64
```

As far as I know, versions of Apple clang less than 503 have this deficiency. Using a newer version of clang is a workaround—either a newer version of Apple clang on newer versions of macOS, or a newer version of llvm.org clang installed manually.

> the problem might not relate to the compiler version but to the OS version

With a little grepping around the SDKs I think I have assured myself that's not the case. `__get_cpuid_max` is an inline function and `__cpuid_count` is a macro, both in a clang header, so it is indeed tied to the compiler version.

> Odd that the toolchain apparently declares them but does not provide them.

Oh, and it doesn't declare them. Earlier in the log:

```
htscodecs/htscodecs/rANS_static4x16pr.c:863:17: warning: implicit declaration of function '__get_cpuid_max' is invalid in C99 [-Wimplicit-function-declaration]
    int level = __get_cpuid_max(0, NULL);
                ^
htscodecs/htscodecs/rANS_static4x16pr.c:864:5: warning: implicit declaration of function '__cpuid_count' is invalid in C99 [-Wimplicit-function-declaration]
    __cpuid_count(0, 0, eax, ebx, ecx, edx);
    ^
```

Here's the [full log on 10.7](https://build.macports.org/builders/ports-10.7_x86_64-builder/builds/184867/steps/install-port/logs/stdio) and for comparison the [log on 10.8](https://build.macports.org/builders/ports-10.8_x86_64-builder/builds/172170/steps/install-port/logs/stdio).

--- a/configure.ac
+++ b/configure.ac
@@ -120,6 +120,25 @@ AS_IF([test "$ax_cv_builtin_prefetch" = "yes"],
            [Define to 1 if you have a __builtin_prefetch])],
 	[])
 
+dnl Check is cpuid works, needed by rANS_static4x16pr.c
+AC_MSG_CHECKING(for cpuid support)
+AC_LINK_IFELSE(
+    [AC_LANG_PROGRAM([
+        #include <cpuid.h>
+        #include <stddef.h>
+    ], [
+        unsigned int eax = 0, ebx = 0, ecx = 0, edx = 0;
+        int level = __get_cpuid_max(0, NULL);
+        if (level >= 1)
+            __cpuid_count(1, 0, eax, ebx, ecx, edx);
+    ])], [
+        AC_MSG_RESULT(yes)
+        AC_DEFINE(HAVE_CPUID, 1, [Set to 1 to use cpuid])
+    ], [
+        AC_MSG_RESULT(no)
+    ]
+)
+
 dnl AC_CHECK_LIB([lzma], [lzma_easy_buffer_encode], [
 dnl 	LIBS="-llzma $LIBS"
 dnl 	AC_DEFINE([HAVE_LIBLZMA],1,[Define to 1 if you have the liblzma library.])])

--- a/configure.ac
+++ b/configure.ac
@@ -120,24 +120,8 @@ AS_IF([test "$ax_cv_builtin_prefetch" = "yes"],
            [Define to 1 if you have a __builtin_prefetch])],
 	[])
 
-dnl Check is cpuid works, needed by rANS_static4x16pr.c
-AC_MSG_CHECKING(for cpuid support)
-AC_LINK_IFELSE(
-    [AC_LANG_PROGRAM([
-        #include <cpuid.h>
-        #include <stddef.h>
-    ], [
-        unsigned int eax = 0, ebx = 0, ecx = 0, edx = 0;
-        int level = __get_cpuid_max(0, NULL);
-        if (level >= 1)
-            __cpuid_count(1, 0, eax, ebx, ecx, edx);
-    ])], [
-        AC_MSG_RESULT(yes)
-        AC_DEFINE(HAVE_CPUID, 1, [Set to 1 to use cpuid])
-    ], [
-        AC_MSG_RESULT(no)
-    ]
-)
+dnl Check is cpuid works, needed by rANS_static4x16pr.c.
+AC_CHECK_DECLS([__get_cpuid_max, __cpuid_count], [], [], [[#include <cpuid.h>]])
 
 dnl AC_CHECK_LIB([lzma], [lzma_easy_buffer_encode], [
 dnl 	LIBS="-llzma $LIBS"

--- a/htscodecs/rANS_static4x16pr.c
+++ b/htscodecs/rANS_static4x16pr.c
@@ -824,7 +824,9 @@ unsigned char *rans_uncompress_O1_4x16(unsigned char *in, unsigned int in_size,
 
 static int rans_cpu = 0xFFFF; // all
 
-#if defined(__x86_64__) && defined(HAVE_CPUID)
+#if defined(__x86_64__) && \
+    defined(HAVE_DECL___CPUID_COUNT)   && HAVE_DECL___CPUID_COUNT && \
+    defined(HAVE_DECL___GET_CPUID_MAX) && HAVE_DECL___GET_CPUID_MAX
 #include <cpuid.h>
 
 #if defined(__clang__) && defined(__has_attribute)

--- a/htscodecs/rANS_static4x16pr.c
+++ b/htscodecs/rANS_static4x16pr.c
@@ -824,8 +824,7 @@ unsigned char *rans_uncompress_O1_4x16(unsigned char *in, unsigned int in_size,
 
 static int rans_cpu = 0xFFFF; // all
 
-#if (defined(__GNUC__) || defined(__clang__)) && defined(__x86_64__)
-// Icc and Clang both also set __GNUC__ on linux, but not on Windows.
+#if defined(__x86_64__) && defined(HAVE_CPUID)
 #include <cpuid.h>
 
 #if defined(__clang__) && defined(__has_attribute)

